src/Orleans/IDs/GrainIdFactory.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using Orleans.Serialization;
+
+namespace Orleans.Runtime
+{
+    internal static class GrainIdFactory
+    {
+        private const int INTERN_CACHE_INITIAL_SIZE = InternerConstants.SIZE_LARGE;
+        private static readonly object lockable = new object();
+        private static readonly TimeSpan internCacheCleanupInterval = InternerConstants.DefaultCacheCleanupFreq;
+        private static Interner<UniqueKey, GrainId> grainIdInternCache;
+
+        public static GrainId NewId()
+        {
+            return FindOrCreateGrainId(UniqueKey.NewKey(Guid.NewGuid(), UniqueKey.Category.Grain));
+        }
+
+        public static GrainId NewClientId(string clusterId = null)
+        {
+            return NewClientId(Guid.NewGuid(), clusterId);
+        }
+
+        internal static GrainId NewClientId(Guid id, string clusterId = null)
+        {
+            return FindOrCreateGrainId(UniqueKey.NewKey(id,
+                clusterId == null ? UniqueKey.Category.Client : UniqueKey.Category.GeoClient, 0, clusterId));
+        }
+
+        internal static GrainId GetGrainId(UniqueKey key)
+        {
+            return FindOrCreateGrainId(key);
+        }
+
+        internal static GrainId GetSystemGrainId(Guid guid)
+        {
+            return FindOrCreateGrainId(UniqueKey.NewKey(guid, UniqueKey.Category.SystemGrain));
+        }
+
+        internal static GrainId GetGrainIdForTesting(Guid guid)
```

Comment:
If we were to have this method (and/or others) as a public utility in the TestingSiloHost project, would it avoid the issue of having way too many InternalsVisibleTo in Abstractions? Again, due to time, we could do that separately, but if it's possible, it might be good to keep in the list of things to change before 2.0